enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,569 @@
+
+We also avoid using `ObjectReference` type for the new fields as its use is [discouraged](https://github.com/kubernetes/api/blob/0cabc089cabafb6be30b40640bbc3ef966a5f00f/core/v1/types.go#L5042-L5058)
+
+The AWS types will not need to be modified, but the use of the `User` field of the `AWSProviderStatus` type will be overloaded to store either the IAM User Name or the IAM Role Name depending on the storage type.
+
+```go
+// AWSStatus containes the status of the credentials request in AWS.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type AWSProviderStatus struct {
+        metav1.TypeMeta `json:",inline"`
+        // User is the name of the User or Role created in AWS for these credentials.
+        User string `json:"user"`
+        // Policy is the name of the policy attached to the user in AWS.
+        Policy string `json:"policy"`
+}
+```
+
+#### `CredentialRequest` Example (Phase 2)
+
+A `CredentialsRequest` CR like this
+```yaml
+apiVersion: cloudcredential.openshift.io/v1
+kind: CredentialsRequest
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: openshift-image-registry
+  namespace: openshift-cloud-credential-operator
+spec:
+  storage:
+    serviceAccount:
+      name: image-registry-sa
+      namespace: openshift-image-registry
+  providerSpec:
+    apiVersion: cloudcredential.openshift.io/v1
+    kind: AWSProviderSpec
+    statementEntries:
+    - effect: Allow
+      action:
+      - "s3:*"
+      resource: "*"
+```
+
+Would result in the creation of an IAM role named
+`sjenning-abcde-openshift-image-registry` with `policy.json`
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "s3:*"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+and a `trust.json`
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "Federated": "arn:aws:iam::111122223333:oidc-provider/s3-us-west-1.amazonaws.com/sjenning-abcde-oidc-provider"
+      },
+      "Action": "sts:AssumeRoleWithWebIdentity",
+      "Condition": {
+        "__doc_comment": "scope the role to the service account (optional)",
+        "StringEquals": {
+          "s3-us-west-1.amazonaws.com/sjenning-abcde-oidc-provider:sub": "system:serviceaccount:openshift-image-registry:image-registry-sa"
+        }
+      }
+    }
+  ]
+}
+
+```
+where `111122223333` is the ARN ID of the OIDC provider.
+
+The controller then creates the `ServiceAccount`
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: image-registry-sa
+  namespace: openshift-image-registry
+  annotations:
+    eks.amazonaws.com/role-arn: "arn:aws:iam::111122223333:role/sjenning-abcde-openshift-image-registry"
+```
+
+Any pod started with this `serviceAccountName` will have the token injected via
```

Comment:
What happens when a Pod is defined that references the not-yet-created Service Account  (CCO hasn't finished creating the role/ServiceAccount)?

Reply:
the webhook does not inject the projected volume.  the annotated SA must exist before the pod is admitted for this to work.  This is a known limitation of the upstream implementation i.e. that a pod can be admitted before the service account that runs it exists.